File: scripts/check_product_sources.py
"""Keep hand-written JavaScript out of the product source trees."""

from pathlib import Path
import subprocess

ROOT = Path(__file__).resolve().parents[1]
JAVASCRIPT_SUFFIXES = {".js", ".mjs", ".cjs", ".jsx"}


def main() -> None:
    listed = subprocess.check_output(
        ["git", "ls-files", "--cached", "--others", "--exclude-standard", "-z", "--", "crates", "local/ui"],
        cwd=ROOT,
    )
    found = sorted(
        Path(name.decode())
        for name in listed.split(b"\0")
        if name and Path(name.decode()).suffix in JAVASCRIPT_SUFFIXES
    )
    if found:
        raise SystemExit("Product JavaScript must be TypeScript source:\n" + "\n".join(map(str, found)))


if __name__ == "__main__":
    main()

File: scripts/check_product_sources.ts
/** Keep hand-written JavaScript out of product source trees. */
import { execFileSync } from 'node:child_process';
import { extname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = fileURLToPath(new URL('..', import.meta.url));
const JAVASCRIPT_SUFFIXES = new Set(['.js', '.mjs', '.cjs', '.jsx']);

export function checkProductSources() {
  const listed = execFileSync(
    'git',
    ['ls-files', '--cached', '--others', '--exclude-standard', '-z', '--', 'crates', 'local/ui'],
    { cwd: ROOT },
  );
  const found = listed
    .toString('utf8')
    .split('\0')
    .filter((name) => name && JAVASCRIPT_SUFFIXES.has(extname(name)))
    .sort();
  if (found.length)
    throw new Error(`Product JavaScript must be TypeScript source:\n${found.join('\n')}`);
}

if (process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  checkProductSources();
}
